_posts/2020-07-27-bikesharing-demand/2020-07-27-bikesharing-demand.Rmd

```diff
@@ -0,0 +1,1175 @@
+    legend.text = element_text(size = 8),
+    legend.title = element_text(size = 9))
+```
+
+To ensure that `regr.ranger.mtry` never exceeds, `importance.filter.nfeat` we
+define a searchspace.
+Since the space for hyperparameters is limited, one could consider here to exploid the whole Parameterspace for nfeat and mtry.
+To save up time we will stick to 28 combinations here, including all possible combinations at 1, 6, 10 and 11 levels.
+
+```{r ranger, message=FALSE, eval=FALSE}
+set.seed(123456)
+ranger_feature = log1p_target_trafo(po_flt)
+at_ranger = AutoTuner$new(ranger_feature,
+  resampling = resampling_cv_5,
+  measures = measures,
+  tune_ps = param_set_ranger,
+  terminator = term("evals", n_evals = 28L),
+  tuner = mlr3tuning::tnr("design_points", design = grid$data)
+)
+# at_ranger$store_tuning_instance = FALSE
+at_ranger$train(bike_task)
+
+archive_ranger = at_ranger$archive("tune_x")[, c("regr.rmse",
+                                                  "importance.filter.nfeat",
+                                                  "regr.ranger.mtry")]
+```
+
+```{r rdsranger, echo=FALSE}
+# saveRDS(archive_ranger, "archive_ranger.rds")
+archive_ranger = readRDS("archive_ranger.rds")
+```
+
+
+In Figure 13 we see the RMSE for the evaluated hyperparameter of the ranger autotuner.
+With increasing `mtry` and the number `features` used in the model, the RMSE decreases.
+The best considered combination is achieved with 11 `features` and
+`mtry = 10`. For higher values of `mtry` and number of features `features` the 
+`RMSE` is slightly higher.
+
+```{r, echo=FALSE, fig.cap="RMSE on tuning grid ranger learner"}
+ggplot(archive_ranger, aes(x = importance.filter.nfeat,
+                           y = regr.ranger.mtry, col = regr.rmse)) +
+  geom_point(size = 3L) +
+  geom_label_repel(
+    data = subset(archive_ranger, regr.rmse == min(regr.rmse)),
+    aes(label = paste(
+      "Min RMSE:", round(regr.rmse, 2L), 
+      "\nfeatures: ",  importance.filter.nfeat, 
+      "\nmtry: ", regr.ranger.mtry)
+    ),
+    size = 2.5,
+    box.padding   = 0.5, 
+    point.padding = 0.5,
+    segment.color = 'grey50',
+    nudge_x = -1.5
+  ) +
+  coord_cartesian(clip = "off") +
+  scale_color_gradientn(colours = heat) +
+  labs(col = "RMSE") +
+  theme_bw() +
+  theme(
+    axis.title = element_text(size = 9),
+    legend.text = element_text(size = 8),
+    legend.title = element_text(size = 9)
+  )
+```
+
+# Benchmarking
+
+Benchmarking has two objectives. First, we want to obtain a performance
+measure for all the fitted models through resampling.
+
+Second, we want to compare the models with each other and examine how tuning
+improved each of them.
+As stated in Chapter 3.3 we use a 3-fold CV as our benchmarking resampling method,
+a 5-fold CV is evaluated inside each autotuning object to obtain the best
+hyperparameter combination.
+Note, that this is not nested resampling, but a step to compare the learners to
```

Comment:
Why is this not nested resampling? Of course it is!

Reply:
From a theoretical perspective benchmarking is model selection...
Theoretically because this point has almost no influence, if the amount of learners compared not `>>1` (like in our case with 3 learners to be compared). In these cases benchmarking GE should be almost equal to nested resampling.
If one compares say 200 learners, the bias in GE would increase, because you do not really have an inner and an outer loop....

Reply:
It is still "nested resampling", the only problem is basically, that the error-estimates after selecting from the benchmark are no longer unbiased. 
(I guess this is what you meant aswell)